platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java

```diff
@@ -170,18 +179,18 @@ public void variants(Action<? super NamedDomainObjectContainer<ConfigurationVari
 
     @Override
     public void capability(Object notation) {
-        if (canCreate) {
-            if (capabilities == null) {
-                capabilities = domainObjectCollectionFactory.newDomainObjectSet(Capability.class);
-            }
-            if (notation instanceof Provider) {
-                capabilities.addLater(((Provider<?>) notation).map(capabilityNotationParser::parseNotation));
-            } else {
-                Capability descriptor = capabilityNotationParser.parseNotation(notation);
-                capabilities.add(descriptor);
-            }
+        if (isObserved()) {
+            throw new InvalidUserCodeException("Cannot declare capability '" + notation + "' on " + displayName + " after " + observationReason.get() + ".");
```

Comment:
💭 If we're going to use `InvalidUserCodeException` instead of `IllegalStateException` here, should we use it for the other failure cases as well?  It's a little odd that the same problem (mutation after observation) produces 2 different exception types depending on the mutation.

Reply:
Switched to `InvalidUserCodeException`